arch/x86/include/ia32/mmustructs.h

```diff
@@ -526,6 +526,19 @@ struct x86_mmu_pt {
  */
 void z_x86_dump_page_tables(struct x86_mmu_pdpt *pdpt);
 
+static inline void z_x86_page_tables_set(struct x86_mmu_pdpt *pdpt)
```

Comment:
Style: I'm not sure an abstracted name is actually helping here.  Setting CR3 is a big event (especially with virtual memory, but even when just changing permissions) and personally I'd prefer to see this called out really clearly and loudly in code rather than being a mild-seeming "set page tables" call.

I mean... doing this from C at all seems like it's likely to be a maintenance danger long term: it really only makes sense with our existing unified physical memory space.  Changing to a VM won't ever make sense with a call like this.   Maybe it should be part of the saved context frame that gets restored by dedicated assembly right before the IRET?  Where else are we calling it?

Reply:
> Style: I'm not sure an abstracted name is actually helping here. Setting CR3 is a big event (especially with virtual memory, but even when just changing permissions) and personally I'd prefer to see this called out really clearly and loudly in code rather than being a mild-seeming "set page tables" call.

I don't agree with you at all. I think instead of putting snippets of assembly code directly in C functions, they should be abstracted to inline functions that do the asm magic. 

MISRA-C agrees with me. See Directive 4.3 Assembly Language shall be encapsulated and isolated.

> I mean... doing this from C at all seems like it's likely to be a maintenance danger long term: it really only makes sense with our existing unified physical memory space. Changing to a VM won't ever make sense with a call like this. 

There's no maintenance danger, this stuff either passes our tests or it doesn't.

> Where else are we calling it?

A CR3 update needs to happen in either two places:

1. If KPTI is disabled, then we update CR3 on context switch. This is where this inline function gets called from z_x86_swap_update_page_tables(), which is invoked from __swap().
2. If KPTI is enabled, then we don't need to update CR3 on context switch; supervisor threads, or threads elevated to privileged mode, all use the global kernel pdpt. Instead, when swap is done and if we trampoline back to user mode via z_x86_trampoline_to_user, CR3 gets set from the kernel's page tables to the per-thread user page tables right before iret.

Reply:
To be clear: I wasn't asking for inline assembly, just for this to be called out in a really obvious way.  It's not a C function, it can't be implemented in C, and in the general case of modified mappings can't even be survived when called from C.

I just don't think "set page tables" properly captures what is happening to a reader.  It didn't to me.

Reply:
I'm not sure then what you want here.

Is your desire that z_x86_swap_update_page_tables() be reimplemented in assembly and place inline with __swap()?

I try, whenever possible, to implement things in C rather than ASM, the compiler is much smarter than I in creating tight asm code.

Reply:
I think if this was called out in code with a big comment to the effect of "This changes CR3, no memory access below is legal unless it is known for sure that the mappings are identical!" (and maybe a similar comment in __swap once it returns to make sure it's touching only the stack) and had a name like "z_x86_set_cr3()", then I'd have nothing to complain about.

Right now there's a comment that mentions CR3, but only in the context of a performance optimization, and I saw that and went "wait, what? This changes CR3?".  Thus the comment here.

Stated differently: right now, _x86_swap_update_page_tables() is piecewise changing mappings that we can reasonably "know" are not related to the current code, where with CR3 you change them all at once.  And as time goes on and we support more stuff like virtual memory, it's going to be changing things into weirder and weirder environments.

Reply:
> I think if this was called out in code with a big comment to the effect of "This changes CR3, no memory access below is legal unless it is known for sure that the mappings are identical!" (and maybe a similar comment in __swap once it returns to make sure it's touching only the stack) and had a name like "z_x86_set_cr3()", then I'd have nothing to complain about.

This is all agreeable to me, will do.

Reply:
Done. This function is now in thread.c, with a barrier and additional comments added.